### DISCUSSION TITLE
How can I test an `EventStream`?

### DISCUSSION THREAD

**Discussion by RemiBardon** [rwf2/Rocket#2804] [Questions]
I have a route which returns an `EventStream`, how can I get the events in tests? 

[`rocket::local::asynchronous::LocalRequest::dispatch`](https://api.rocket.rs/v0.5/rocket/local/asynchronous/struct.LocalRequest#method.dispatch) returns a [`LocalResponse`](https://api.rocket.rs/v0.5/rocket/local/asynchronous/struct.LocalResponse) and I see no way to read events from it. Getting the response body returns an empty string, and there is no other API to get the stream AFAIK.

~~I tried sleeping before reading the body, it doesn't work either.~~
Edit: I tried yielding a static event in the `EventStream` and I see it in the response body. I can probably decode a list of events then.

**Comment by the10thWiz** [comment]
Yes. You will want to take advantage of [`Body`](https://api.rocket.rs/v0.5/rocket/response/struct.Body#impl-AsyncRead-for-Body%3C'_%3E)'s impl of `AsyncRead`. You can use [`AsyncReadExt`](https://docs.rs/tokio/1.38.0/tokio/io/trait.AsyncReadExt.html) (`rocket::tokio::io::AsyncReadExt`) to read raw bytes (and decode to a list of events), or `Body`'s `into_string` or `into_bytes` to read the full data.

If you didn't use `rocket::tokio::time::sleep` to sleep, the request may not be able to make any progress. This would have prevented the handler from sending any events.

**Reply by RemiBardon** [reply]
I could read the body but now I'm trying to figure out how to parse the events.

Also yes I used `rocket::tokio::time::sleep` to make sure the thread is not blocked. I wasn't receiving any event because of a `panic!()` inside the `EventStream!` which wasn't logged 🤦🏻‍♂️

**Reply by SergioBenitez** [reply]
> Also yes I used rocket::tokio::time::sleep to make sure the thread is not blocked. I wasn't receiving any event because of a panic!() inside the EventStream! which wasn't logged 🤦🏻‍♂️

That sounds odd. A panic is logged at a very low level, so it would be quite peculiar if that were the case. Do you have a reproducible example?

**Reply by RemiBardon** [reply]
No I don't and honestly it doesn't surprise me. I'm using cucumber to run BDD tests and it doesn't integrate well with Rocket's logging and system logs. I still have to improve this one day, it's quite annoying.

**Reply by the10thWiz** [reply]
> I'm trying to figure out how to parse the events.

Obviously, it depends on what format the events are in, but `serde_json` provides a [`StreamDeserializer`](https://docs.rs/serde_json/latest/serde_json/struct.StreamDeserializer.html) that deserializes a stream of multiple JSON objects. I think this might be an opportunity to improve Rocket's testing functionality, and provide methods to stream events from the async local response.

**Reply by the10thWiz** [reply]
> That sounds odd. A panic is logged at a very low level, so it would be quite peculiar if that were the case. Do you have a reproducible example?

Panics inside `EventStream` are not caught (or logged) by Rocket, so they would presumably be logged to stderr, while Rocket normally logs to stdout. If the framework (cucumber) doesn't track stderr, the panic would be lost.
